### PR TITLE
branch node

### PR DESCRIPTION
**Pull request by surindersinghp**

### CODE REVIEW

File: pkg/controller/nodes/branch/handler.go [hunk 1]
```diff
@@ -18,19 +22,135 @@ func (b *branchHandler) FinalizeRequired() bool {
...
+
+		var ok bool
+		finalNode, ok := w.GetNode(*finalNodeId)
+		if !ok {
+			errMsg := fmt.Sprintf("Branch downstream finalized node not found. FinalizedNode [%s]", *finalNodeId)
+			logger.Debugf(ctx, errMsg)
+			return handler.DoTransitionToFailed(errors.DownstreamNodeNotFoundError, errMsg), nil
+		}
+		i := nCtx.NodeID()
+		childNodeStatus := w.GetNodeExecutionStatus(finalNode.GetID())
```
Comment (surindersinghp) [inline pkg/controller/nodes/branch/handler.go]: @kumare this should be moved to branchstate but in next PR. lets go with this one in this PR

File: pkg/controller/nodes/branch/handler.go [hunk 2]
```diff
@@ -18,19 +22,135 @@ func (b *branchHandler) FinalizeRequired() bool {
...
-func (b *branchHandler) Handle(ctx context.Context, executionContext handler.NodeExecutionContext) (handler.Transition, error) {
-	panic("implement me")
+func (b *branchHandler) Handle(ctx context.Context, nCtx handler.NodeExecutionContext) (handler.Transition, error) {
+	logger.Debug(ctx, "Starting Branch Node")
+	branchNode := nCtx.Node().GetBranchNode()
+	if branchNode == nil {
+		return handler.DoTransitionToFailed(errors.IllegalStateError, "Invoked branch handler, for a non branch node."), nil
+	}
+
+	w := nCtx.Workflow()
```
Comment (kumare3) [inline pkg/controller/nodes/branch/handler.go]: do at the last minute

File: pkg/controller/nodes/branch/handler.go [hunk 3]
```diff
@@ -18,19 +22,135 @@ func (b *branchHandler) FinalizeRequired() bool {
...
+	nodeInputs, err := nCtx.InputReader().Get(ctx)
```
Comment (kumare3) [inline pkg/controller/nodes/branch/handler.go]: lets read inputs in the if block

File: pkg/controller/nodes/branch/handler.go [hunk 4]
```diff
@@ -18,19 +22,135 @@ func (b *branchHandler) FinalizeRequired() bool {
...
+	if err != nil {
+		errMsg := fmt.Sprintf("Failed to read input. Error [%s]", err)
+		return handler.DoTransitionToFailed(errors.RuntimeExecutionError, errMsg), nil
+	}
+
+	if nCtx.NodeStatus().GetPhase() == v1alpha1.NodePhaseNotYetStarted {
```
Comment (kumare3) [inline pkg/controller/nodes/branch/handler.go]: we should look at the branch state

File: pkg/controller/nodes/branch/handler.go [hunk 5]
```diff
@@ -18,19 +22,135 @@ func (b *branchHandler) FinalizeRequired() bool {
...
+		childNodeStatus.SetParentNodeID(&i)
+
+		logger.Debugf(ctx, "Recursing down branchNodestatus node")
+		nodeStatus := w.GetNodeExecutionStatus(nCtx.NodeID())
+		return b.recurseDownstream(ctx, nCtx, nodeStatus, finalNode)
+	}
+
+	// If the branchNodestatus was already evaluated i.e, Node is in Running status
+	nodeStatus := w.GetNodeExecutionStatus(nCtx.NodeID())
+	branchStatus := nodeStatus.GetOrCreateBranchStatus()
```
Comment (kumare3) [inline pkg/controller/nodes/branch/handler.go]: lets not access branchStatus, you can read it from NodeState, not accessing the status will make it easier to get rid of the status

File: pkg/controller/nodes/branch/handler.go [hunk 6]
```diff
@@ -18,19 +22,135 @@ func (b *branchHandler) FinalizeRequired() bool {
...
+		// For branch node we set the output node to be the same as the child nodes output
+		childNodeStatus := w.GetNodeExecutionStatus(branchTakenNode.GetID())
+		nodeStatus.SetDataDir(childNodeStatus.GetDataDir())
+		phase := handler.PhaseInfoSuccess(nil)
+		return handler.DoTransition(handler.TransitionTypeEphemeral, phase), nil
+	}
+
+	if downstreamStatus.HasFailed() {
+		errMsg := downstreamStatus.Err.Error()
+		code, _ := errors.GetErrorCode(downstreamStatus.Err)
```
Comment (kumare3) [inline pkg/controller/nodes/branch/handler.go]: if errcode not found, then we should be "unknown"

File: pkg/controller/nodes/branch/handler.go [hunk 7]
```diff
@@ -18,19 +22,135 @@ func (b *branchHandler) FinalizeRequired() bool {
...
+	phase := handler.PhaseInfoRunning(nil)
+	return handler.DoTransition(handler.TransitionTypeEphemeral, phase), nil
 }
 
-func (b *branchHandler) Abort(ctx context.Context, executionContext handler.NodeExecutionContext) error {
-	panic("implement me")
+func (b *branchHandler) Abort(ctx context.Context, nCtx handler.NodeExecutionContext) error {
+
+	branch := nCtx.Node().GetBranchNode()
+	w := nCtx.Workflow()
```
Comment (kumare3) [inline pkg/controller/nodes/branch/handler.go]: later

File: pkg/controller/nodes/branch/handler.go [hunk 8]
```diff
@@ -18,19 +22,135 @@ func (b *branchHandler) FinalizeRequired() bool {
...
+	if branch == nil {
+		return errors.Errorf(errors.IllegalStateError, w.GetID(), nCtx.NodeID(), "Invoked branch handler, for a non branch node.")
+	}
+
+	// If the branch was already evaluated i.e, Node is in Running status
```
Comment (kumare3) [inline pkg/controller/nodes/branch/handler.go]: look at branchstate and decide if you want to abort downstream

File: pkg/controller/nodes/branch/handler.go [hunk 9]
```diff
@@ -18,19 +22,135 @@ func (b *branchHandler) FinalizeRequired() bool {
 }
```
Comment (kumare3) [inline pkg/controller/nodes/branch/handler.go]: false

File: pkg/controller/nodes/handler/transition.go [hunk 10]
```diff
@@ -30,3 +32,9 @@ var UnknownTransition = Transition{TransitionTypeEphemeral, PhaseInfoUndefined}
 func DoTransition(ttype TransitionType, info PhaseInfo) Transition {
 	return Transition{ttype: ttype, info: info}
 }
+
+func DoTransitionToFailed(code errors.ErrorCode, errorMsg string) Transition {
```
Comment (kumare3) [inline pkg/controller/nodes/handler/transition.go]: get rid of this method?

**Pull request merged**